proposals/required-members.md

```diff
@@ -263,42 +263,131 @@ init_argument_initializer
+    protected Base() { }
+}
+public class Derived : Base, I
+{
+    required int I.Prop1 { get; } // Error: explicit interface implementions cannot be required as they cannot be set in an object initializer
+
+    public required override int Prop2 { get; } // Error: this property is hidden by Derived.Prop2 and cannot be set in an object initializer
+    public new int Prop2 { get; }
+}
+```
+
+It is an error to hide a `required` member, as that member can no longer be set by a consumer.
+
+When overriding a `required` member, the `required` keyword must be included on the method signature. This is done so that if we ever want to allow
+unrequiring a property with an override in the future, we have design space to do so.
+
+Overrides are allowed to mark a member `required` where it was not `required` in the base type. A member so-marked is added to the required members
+list of derived type.
+
+### Metadata Representation
+
+The following 2 attributes are known to the C# compiler and required for this feature to function:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = false)]
+public sealed class RequiredMembersAttribute : Attribute
+{
+    public RequiredMembersAttribute(string[] members) => Members = members;
+    public string[] Members { get; }
+}
+
+[AttributeUsage(AttributeTargets.Constructor)]
+public sealed class NoRequiredMembersAttribute : Attribute
+{
+    public NoRequiredMembersAttribute() {}
+}
+```
+
+It is an error to manually apply `RequiredMembersAttribute` to a type.
+
+All members that are required by a type are listed in a `RequiredMembersAttribute` on the type. Members from any base type are not included, except
+when an override adds `required` to a member that was not `required` in the base type. As these members can only be fields or properties, duplicate
+names cannot exist.
+
+Any constructor in a type with `required` members that does not have `NoRequiredMembers` applied to it is marked with an `Obsolete` attribute with
+the string `"Types with required members are not supported in this version of your compiler"`, and the attribute is marked as an error, to prevent
+any older compilers from using these constructors. We don't use a `modreq` here because it is a goal to maintain binary compat: if the last `required`
+property was removed from a type, the compiler would no longer synthesize this `modreq`, which is a binary-breaking change and all consumers would
+need to be recompiled. A compiler that understands `required` members will ignore this obsolete attribute. Note that members can come from base types
+as well: even there are no new `required` members in the current type, if any base type has `required` members, this `Obsolete` attribute will be
+generated. If the constructor already has an `Obsolete` attribute, no additional attribute will be generated. If a by-ref-like type also has `required`
```

Comment:
>If the constructor already has an `Obsolete` attribute, no additional attribute will be generated. [](http://example.com/codeflow?start=11&length=98)

Is that what we do for other "not supported in this version of your compiler" Obsolete attributes? What if the explicit Obsolete attribute is only marked as "warning"?

Reply:
Yes, it is. If a type has an obsolete marker, it takes precedence, no matter the warning level. This is even true for `ref struct` types: https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Source/SourceNamedTypeSymbol.cs,1544